File: 0502-ipo/0502-ipo.java
import java.util.*;

public class Solution {
    public int findMaximizedCapital(int k, int w, int[] profits, int[] capital) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        List<int[]> temp = new ArrayList<>();
        
        for (int i = 0; i < profits.length; i++) {
            temp.add(new int[]{capital[i], profits[i]});
        }
        
        temp.sort(Comparator.comparingInt(a -> a[0]));
        
        for (int i = 0; i < temp.size(); i++) {
            if (temp.get(i)[0] <= w) {
                pq.add(temp.get(i)[1]);
            } else {
                if (pq.isEmpty()) return w;
                w += pq.poll();
                k--;
                
                while (!pq.isEmpty() && w < temp.get(i)[0] && k > 0) {
                    w += pq.poll();
                    k--;
                }
                
                if (k == 0) return w;
                
                if (w >= temp.get(i)[0]) {
                    pq.add(temp.get(i)[1]);
                }
            }
            if (k == 0) break;
        }
        
        while (!pq.isEmpty() && k > 0) {
            w += pq.poll();
            k--;
        }
        
        return w;
    }
}

File: 0502-ipo/0502-ipo.cpp
class Solution {
public:
    int findMaximizedCapital(int k, int w, vector<int>& profits, vector<int>& capital) {
        priority_queue<long long> pq; 
        vector<pair<int,int>> temp;
        for(int i=0;i<profits.size();i++){
            temp.push_back({capital[i],profits[i]});
        }
        sort(temp.begin(),temp.end());
        for(int i=0;i<temp.size();i++){
            if(temp[i].first<=w){
                pq.push(temp[i].second);
            }else{
                if(pq.empty())return w;
                while(!pq.empty() && w<temp[i].first && k>0){
                    w+=pq.top();
                    pq.pop();
                    k--;
                }
                if(k==0)return w;
                if(w>=temp[i].first)
                pq.push(temp[i].second);
            }
            if(k==0)break;
        }
        while(!pq.empty() && k>0){
            w+=pq.top();
            pq.pop();
            k--;
        }
        return w;
    }
};